src/rgw/rgw_auth_s3.h

```diff
@@ -309,50 +326,92 @@ class AWSv4ComplMulti : public rgw::auth::Completer,
       return signature;
     }
 
+    size_t get_offset() { return data_offset_in_stream; }
+
     /* Factory: create an object representing metadata of first, initial chunk
      * in a stream. */
-    static ChunkMeta create_first(const std::string_view& seed_signature) {
-      return ChunkMeta(seed_signature);
+    static ChunkMeta create_first(AWSv4ComplMulti* completer,
+				  const std::string_view& seed_signature,
+				  uint32_t flags,
+				  uint32_t cnt) {
+      return ChunkMeta(completer, seed_signature, flags, cnt);
     }
 
     /* Factory: parse a block of META_MAX_SIZE bytes and creates an object
      * representing non-first chunk in a stream. As the process is sequential
      * and depends on the previous chunk, caller must pass it. */
-    static std::pair<ChunkMeta, size_t> create_next(CephContext* cct,
+    static std::pair<ChunkMeta, size_t> create_next(AWSv4ComplMulti* completer,
+						    CephContext* cct,
                                                     ChunkMeta&& prev,
                                                     const char* metabuf,
-                                                    size_t metabuf_len);
+                                                    size_t metabuf_len,
+						    uint32_t flags);
   } chunk_meta;
 
   size_t stream_pos;
+  uint32_t content_length;
   boost::container::static_vector<char, ChunkMeta::META_MAX_SIZE> parsing_buf;
+
+  /* XXX future optimization: combined into one 256-byte vector */
+  static constexpr size_t trailer_buf_size = 256;
+  boost::container::static_vector<char, trailer_buf_size> trailer_vec;
```

Comment:
the purpose in radoslaw's using static_vector is to avoid allocations, and I don't really see a need to revise that

Reply:
understood, but in the common case small_vector wouldn't allocate either. the difference is just how we treat the unexpected requests that overrun - either we allocate some memory to handle them correctly, or return an error to reject requests which are otherwise valid

Reply:
I think we should put an upper limit on how large we allow this to grow, but I guess this makes sense

Reply:
I've moved all of this logic to AWSv4ComplMulti::complete(), which does streamline the code.  I think it's pretty clear from both AWS "doc" and client SDK code comments that AWS implementers don't intend to support arbitrary trailing headers, though, at least for now.  I think that argues for keeping the static_vector for recognizing trailers.

Reply:
ok, agreed that we'd need to pick some arbitrary limit even if we did support growth

however, i think it's worth enforcing this limit with an explicit error. as-is, it looks like we'll read up to 256 bytes and ignore anything else

in the case of STREAMING-AWS4-HMAC-SHA256-PAYLOAD-TRAILER, a truncated read is likely to end up failing to parse/validate the trailer signature. but for STREAMING-UNSIGNED-PAYLOAD-TRAILER, it looks like we'd succeed but silently omit some of the headers from x-amz-trailer. omitting trailing headers like x-amz-checksum-* would prevent us from performing the requested integrity check